### ISSUE TITLE
Improved cron trigger to allow parallel workflows

### ISSUE THREAD

**Issue by brendanheywood** [catalyst/moodle-tool_dataflows#169]
Builds on https://github.com/catalyst/moodle-tool_dataflows/issues/94

* [x] if there is more than one workflow to be run in a given minute, spawn adhoc tasks for all the others except one
* [x] run one of them right now so we don't add more latency
* [x] introduce locking for each workflow so only once instance can run at a given point

**Comment by brendanheywood** [comment]
So I'm going to refine the requirements here a bit which summarized the chat in matrix earlier:

1) Long term we want the ability for some workflows to be able to run at the same time as other instances of the same type of workflow, eg if a webhook fires or moodle event triggers a dataflow then two instances will be created one after the other and both run just fine.
2) If a flow should or should not run concurrently this should be declared by the trigger step
3) moment we only have a single trigger type of cron, so its a little harder to see examples of the other types
4) Both to make it easier to test, and because we will actually want it in production lets make the 'concurrency allowed' an option in the cron trigger step

So I think this boils down to tasks like:
* [ ] add config at the dataflow level will says whether or not concurrency is allowed
* [ ] specifically in the cron trigger, if it allows concurrency then each new run will spawn a new via queue_adhoc_task and that ad hoc task will just run and there isn't any extra locking
* [ ] if it doesn't allow concurrency then still spawn an adhoc task, but spawn a new task via reschedule_or_queue_adhoc_task. This will help to avoid two adhoc tasks for the same flow colliding and reduce but completely eliminate the need for locks, but we'll still lock either way
* [ ] when we lock I think just do a 0 timeout, if it can't get a lock then we know it is running so just exit and the cron api will faildelay it and rerun it again later
* [ ] Various steps will be incompatible with concurrency and we should validate for this. The one example I can think of right now is that using the sql reader Counter field will be incompatible with concurrency as it will create race conditions, so add a validation rule that if it is set AND the flow concurrency is allowed
* [ ] as an extra test, even running a flow manually should not be allowed if concurrency is off and you run it twice in two tabs or two sessions

**Comment by jaypha** [comment]
Currently, if the dataflow is blocked, it displays this.
![Screenshot from 2022-07-26 17-13-39](https://user-images.githubusercontent.com/1052990/180946159-21eb72c9-c5cc-4e55-9e17-7b90718d8463.png)
Do you want it to display something in addition to this?

**Comment by jaypha** [comment]
The cron actually never fails. Everything is wrapped inside try-catch blocks. Do you still want the blocked dataflows to just exit?

**Comment by brendanheywood** [comment]
Yes if it is already running then exit immediately.

One nice thing is when you lock a workflow we stash some metadata somewhere around who and what is running it and then we can display that when it is locked eg 'This workflow is already being run by cron started at 12:45 pid 12345' or 'This workflow is already being run by Brendan Heywood started at 12:45'
